packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts

```diff
@@ -79,29 +80,49 @@ export default createRule<Options, MessageIds>({
         ) &&
         // ignore class properties as they are compile time guarded
         // also ignore function arguments as they can't be used before defined
-        ts.isVariableDeclaration(declaration) &&
-        // is it `const x!: number`
-        declaration.initializer === undefined &&
-        declaration.exclamationToken === undefined &&
-        declaration.type !== undefined
+        ts.isVariableDeclaration(declaration)
       ) {
-        // check if the defined variable type has changed since assignment
-        const declarationType = checker.getTypeFromTypeNode(declaration.type);
-        const type = getConstrainedTypeAtLocation(services, node);
         if (
```

Comment:
(nit, optional) i think a comment would be helpful here, since it's pretty confusing territory. Maybe something like this?

```text
For var declarations, we need to check whether the node 
is actually in a descendant of its declaration or not. If not, 
it may be used before defined.

eg 
if (Math.random() < 0.5) {
    var x: number  = 2;
} else {
    x!.toFixed();
}
```